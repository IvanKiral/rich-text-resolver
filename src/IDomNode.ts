export interface IDomNode {
    type: 'text' | 'tag'
}

export interface IDomTextNode extends IDomNode {
    content: string
}

export interface IDomHtmlNode extends IDomNode{
    tagName: string,
    attributes: Record<string, string>,
    children: IDomNode[]
}