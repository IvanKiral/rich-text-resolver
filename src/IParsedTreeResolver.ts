import { IParsedTree } from "./IRichTextToJsonTransformer";

export interface IParsedTreeResolver {
    resolve(parsedTree: IParsedTree): any;
}