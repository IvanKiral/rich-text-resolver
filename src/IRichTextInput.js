"use strict";
// TODO zkontrolovat optional a required
Object.defineProperty(exports, "__esModule", { value: true });
