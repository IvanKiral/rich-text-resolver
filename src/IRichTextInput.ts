// TODO zkontrolovat optional a required

export type IRichTextInput = {
    value: string,
    images: {[key: string]: {
        image_id: string,
        description: string,
        url: string,
        width: number,
        height: number
    }},
    links: {[key: string]: {
        codename: string,
        type: string,
        url_slug: string
    }},
    modular_content: string[]
}