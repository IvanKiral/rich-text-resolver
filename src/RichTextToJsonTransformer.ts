import { Elements } from "@kontent-ai/delivery-sdk";
import { IHtmlNode, IRichTextToJsonTransformer, IParsedTree, IParserHtmlNode, IParserNode, IParserTextNode } from "./IRichTextToJsonTransformer";
import { parse, HTMLElement, Node, NodeType, TextNode } from 'node-html-parser';


class Transformer { //implements IRichTextToJsonTransformer {
    transform(dummyRichText: Elements.RichTextElement): IHtmlNode[] {
        const parsedHtml = parse(dummyRichText.value);
        parsedHtml.childNodes
        return this.remapNode(parsedHtml);
    }
    
    traverse(parsedRichText: Node): IParserNode[] {
        var transformedTree: IParsedTree = {
            content: []
        };

        var transformedNodes: IParserNode[] = [];

        if (parsedRichText.childNodes.length > 0 || parsedRichText.nodeType) {
            switch (parsedRichText.nodeType) {
                case 1: {
                    let htmlNode: IParserHtmlNode = {
                        name: (<HTMLElement>parsedRichText).rawTagName,
                        attributes: (<HTMLElement>parsedRichText).attributes,
                        children: (<HTMLElement>parsedRichText).childNodes.flatMap((childNode: Node) => this.traverse(childNode)),
                        type: 'tag'
                    }
                    transformedNodes.push(htmlNode);
                    break;
                }

                case 3: {
                    let textNode: IParserTextNode = {
                        content: (<TextNode>parsedRichText).rawText,
                        type: 'text'
                    }
                    transformedNodes.push(textNode);
                }
                default:
                    break;
            }
        }

        return transformedNodes;
    }

    parseInternal(richTextElement: Elements.RichTextElement): HTMLElement {
        return parse(richTextElement.value);
    }

    parse(dummyRichText: Elements.RichTextElement): HTMLElement {
        var parsedTree: IParserNode[] = [];
        const parseResult = parse(dummyRichText.value);

        return parseResult;
    }

    remapNode(parsedHtml: HTMLElement): IHtmlNode[] {
        var nodes: IHtmlNode[] = [];
        if(parsedHtml.childNodes.length > 0 || parsedHtml.rawTagName) {            
            nodes.push({
                tagName: parsedHtml.rawTagName ?? "",
                attributes: parsedHtml.attributes ?? {},
                innerContent: parsedHtml.innerHTML ?? '',
                children: parsedHtml.childNodes.flatMap((element) => this.remapNode(element as HTMLElement))
            });
        }
        return nodes;
    }
}

export const transformer = new Transformer();