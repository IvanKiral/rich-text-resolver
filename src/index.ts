export * from './parser';
export * from './resolver';
export * from './utils';