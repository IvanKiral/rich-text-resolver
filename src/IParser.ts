import { IDomNode } from "./IDomNode"

export interface IParser<TInput> {
    parse(value: TInput): {
        children: IDomNode[]
    }
}