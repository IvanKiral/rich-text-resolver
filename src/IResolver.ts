import { IDomNode } from "./IDomNode";

export interface IOutputResult<TOutput> {
    reolvedNode: TOutput,
    nodes: TOutput[]
}

export interface IResolver<TInput, TOutput> {
    resolveAsync(input: TInput, resolvers: {
        resolveDomNode(domNode: IDomNode): Promise<IOutputResult<TOutput>>
    })
}