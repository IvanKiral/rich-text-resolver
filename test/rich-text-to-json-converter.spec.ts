import { Elements, ElementType } from "@kentico/kontent-delivery";
import { transformer } from "../src/RichTextToJsonTransformer";

describe("rich-text-to-json-converter", () => {

    it("empty rich text converts properly", () => {
        const dummyRichText: Elements.RichTextElement = {
            value: "<p><br></p>",
            type: ElementType.RichText,
            images: [],
            linkedItemCodenames: [],
            linkedItems: [],
            links: [],
            name: "dummy"
        };

        const result: any = transformer.transform(dummyRichText);

        expect(result).toEqual<any>([
            {
                tagName: "p",
                attributes: [],
                textContent: "",
                children: [
                    {
                        tagName: "br",
                        attributes: [],
                        children: []
                    }
                ],
            }
        ]);

    });
});
