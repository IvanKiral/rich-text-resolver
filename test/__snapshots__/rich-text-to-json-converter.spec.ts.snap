// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new rich text parser parses rich text 1`] = `
Object {
  "children": Array [
    Object {
      "attributes": Object {
        "class": "test",
        "id": "3",
      },
      "children": Array [
        Object {
          "attributes": Object {
            "data-codename": "test_item",
            "data-rel": "component",
            "data-type": "item",
            "type": "application/kenticocloud",
          },
          "children": Array [],
          "tagName": "object",
          "type": "tag",
        },
        Object {
          "content": "before text",
          "type": "text",
        },
        Object {
          "attributes": Object {
            "href": "mailto:email@abc.test",
          },
          "children": Array [
            Object {
              "content": "email",
              "type": "text",
            },
          ],
          "tagName": "a",
          "type": "tag",
        },
        Object {
          "content": "after text line break ",
          "type": "text",
        },
        Object {
          "attributes": Object {},
          "children": Array [],
          "tagName": "br",
          "type": "tag",
        },
      ],
      "tagName": "p",
      "type": "tag",
    },
    Object {
      "attributes": Object {},
      "children": Array [
        Object {
          "content": "text span",
          "type": "text",
        },
      ],
      "tagName": "span",
      "type": "tag",
    },
  ],
}
`;
